File: src/features/scheduler/domain/swap.ts
import { DaySchedule, ShiftSlotRef } from "./types";

function slotIds(row: DaySchedule, shiftType: ShiftSlotRef["shiftType"]) {
  return shiftType === "night" ? row.nightIds : row.dayIds;
}

function slotNames(row: DaySchedule, shiftType: ShiftSlotRef["shiftType"]) {
  return shiftType === "night" ? row.night : row.day;
}

export function isSameSlot(left: ShiftSlotRef, right: ShiftSlotRef) {
  return left.date === right.date && left.shiftType === right.shiftType && left.slot === right.slot;
}

export function findAssignedMemberId(assignments: DaySchedule[], ref: ShiftSlotRef): string | null {
  const row = assignments.find((assignment) => assignment.date === ref.date);
  if (!row) return null;
  return slotIds(row, ref.shiftType)[ref.slot] ?? null;
}

export function validateSwapProposal(input: {
  assignments: DaySchedule[];
  requesterId: string;
  requesteeId: string;
  requesterShift: ShiftSlotRef;
  requesteeShift: ShiftSlotRef;
  now?: number;
}) {
  if (input.requesterId === input.requesteeId) {
    throw new Error("You cannot propose a swap with yourself");
  }

  if (isSameSlot(input.requesterShift, input.requesteeShift)) {
    throw new Error("Choose two different shifts to swap");
  }

  const today = new Date(input.now ?? Date.now()).toISOString().slice(0, 10);
  for (const ref of [input.requesterShift, input.requesteeShift]) {
    if (!/^\d{4}-\d{2}-\d{2}$/.test(ref.date)) {
      throw new Error(`Invalid date: ${ref.date}`);
    }
    if (ref.date < today) {
      throw new Error(`Cannot propose a swap for ${ref.date} because that date has already passed`);
    }
  }

  if (findAssignedMemberId(input.assignments, input.requesterShift) !== input.requesterId) {
    throw new Error("You are not assigned to the shift you want to give up");
  }

  if (findAssignedMemberId(input.assignments, input.requesteeShift) !== input.requesteeId) {
    throw new Error("The selected teammate is not assigned to that shift");
  }
}

function withSlotValue(
  row: DaySchedule,
  ref: ShiftSlotRef,
  value: { id: string | null; name: string | null },
): DaySchedule {
  const nextIds = [...slotIds(row, ref.shiftType)] as [string | null, string | null];
  const nextNames = [...slotNames(row, ref.shiftType)] as [string | null, string | null];
  nextIds[ref.slot] = value.id;
  nextNames[ref.slot] = value.name;

  return ref.shiftType === "night"
    ? { ...row, nightIds: nextIds, night: nextNames }
    : { ...row, dayIds: nextIds, day: nextNames };
}

export function applySwap(
  assignments: DaySchedule[],
  requesterShift: ShiftSlotRef,
  requesteeShift: ShiftSlotRef,
): DaySchedule[] {
  const requesterRow = assignments.find((row) => row.date === requesterShift.date);
  const requesteeRow = assignments.find((row) => row.date === requesteeShift.date);
  if (!requesterRow || !requesteeRow) {
    throw new Error("The shifts being swapped no longer exist in the schedule");
  }

  const requesterValue = {
    id: slotIds(requesterRow, requesterShift.shiftType)[requesterShift.slot] ?? null,
    name: slotNames(requesterRow, requesterShift.shiftType)[requesterShift.slot] ?? null,
  };
  const requesteeValue = {
    id: slotIds(requesteeRow, requesteeShift.shiftType)[requesteeShift.slot] ?? null,
    name: slotNames(requesteeRow, requesteeShift.shiftType)[requesteeShift.slot] ?? null,
  };

  return assignments.map((row) => {
    let next = row;
    if (row.date === requesterShift.date) {
      next = withSlotValue(next, requesterShift, requesteeValue);
    }
    if (row.date === requesteeShift.date) {
      next = withSlotValue(next, requesteeShift, requesterValue);
    }
    return next;
  });
}

File: backend/app/domain/swap.py
from __future__ import annotations
from datetime import datetime, timezone

from app.domain.types import DaySchedule, ShiftSlotRef


def _slot_ids(row: DaySchedule, shift_type: str) -> tuple[str | None, str | None]:
    return row.night_ids if shift_type == "night" else row.day_ids


def _slot_names(row: DaySchedule, shift_type: str) -> tuple[str | None, str | None]:
    return row.night if shift_type == "night" else row.day


def is_same_slot(left: ShiftSlotRef, right: ShiftSlotRef) -> bool:
    return left.date == right.date and left.shift_type == right.shift_type and left.slot == right.slot


def find_assigned_member_id(assignments: list[DaySchedule], ref: ShiftSlotRef) -> str | None:
    for row in assignments:
        if row.date == ref.date:
            ids = _slot_ids(row, ref.shift_type)
            return ids[ref.slot]
    return None


def validate_swap_proposal(
    assignments: list[DaySchedule],
    requester_id: str,
    requestee_id: str,
    requester_shift: ShiftSlotRef,
    requestee_shift: ShiftSlotRef,
    now_ms: int | None = None,
) -> None:
    if requester_id == requestee_id:
        raise ValueError("You cannot propose a swap with yourself")

    if is_same_slot(requester_shift, requestee_shift):
        raise ValueError("Choose two different shifts to swap")

    now = datetime.fromtimestamp((now_ms or int(datetime.now(timezone.utc).timestamp() * 1000)) / 1000, tz=timezone.utc)
    today = now.date().isoformat()

    for ref in (requester_shift, requestee_shift):
        if ref.date < today:
            raise ValueError(f"Cannot propose a swap for {ref.date} because that date has already passed")

    if find_assigned_member_id(assignments, requester_shift) != requester_id:
        raise ValueError("You are not assigned to the shift you want to give up")

    if find_assigned_member_id(assignments, requestee_shift) != requestee_id:
        raise ValueError("The selected teammate is not assigned to that shift")


def _with_slot_value(
    row: DaySchedule,
    ref: ShiftSlotRef,
    new_id: str | None,
    new_name: str | None,
) -> DaySchedule:
    ids = list(_slot_ids(row, ref.shift_type))
    names = list(_slot_names(row, ref.shift_type))
    ids[ref.slot] = new_id
    names[ref.slot] = new_name
    id_pair: tuple[str | None, str | None] = (ids[0], ids[1])
    name_pair: tuple[str | None, str | None] = (names[0], names[1])

    if ref.shift_type == "night":
        return DaySchedule(row.date, row.weekday_label, row.weekend, name_pair, row.day, id_pair, row.day_ids)
    else:
        return DaySchedule(row.date, row.weekday_label, row.weekend, row.night, name_pair, row.night_ids, id_pair)


def apply_swap(
    assignments: list[DaySchedule],
    requester_shift: ShiftSlotRef,
    requestee_shift: ShiftSlotRef,
) -> list[DaySchedule]:
    requester_row = next((r for r in assignments if r.date == requester_shift.date), None)
    requestee_row = next((r for r in assignments if r.date == requestee_shift.date), None)

    if not requester_row or not requestee_row:
        raise ValueError("The shifts being swapped no longer exist in the schedule")

    requester_id = _slot_ids(requester_row, requester_shift.shift_type)[requester_shift.slot]
    requester_name = _slot_names(requester_row, requester_shift.shift_type)[requester_shift.slot]
    requestee_id = _slot_ids(requestee_row, requestee_shift.shift_type)[requestee_shift.slot]
    requestee_name = _slot_names(requestee_row, requestee_shift.shift_type)[requestee_shift.slot]

    result: list[DaySchedule] = []
    for row in assignments:
        next_row = row
        if row.date == requester_shift.date:
            next_row = _with_slot_value(next_row, requester_shift, requestee_id, requestee_name)
        if row.date == requestee_shift.date:
            next_row = _with_slot_value(next_row, requestee_shift, requester_id, requester_name)
        result.append(next_row)

    return result
